### PR TITLE
Handle default namespace in config-updater

### PR DESCRIPTION
**Pull request by hongkailiu**
Fix https://github.com/kubernetes/test-infra/issues/15570

/cc @stevekuznetsov @petr-muller @cjwagner

### CODE REVIEW

**Comment by hongkailiu** [comment]
/hold
working on unit test for the case

**Comment by hongkailiu** [comment]
> Can we get some unit tests that fail before this patch and succeed after it?

I knew this is coming. Working on it right now.

File: prow/plugins/updateconfig/updateconfig.go [hunk 1]
```diff
@@ -240,6 +240,22 @@ func FilterChanges(cfg plugins.ConfigUpdater, changes []github.PullRequestChange
 			}
 		}
 	}
+	return handleDefaultNamespace(toUpdate, defaultNamespace)
+}
+
+// handleDefaultNamespace ensures plugins.ConfigMapID.Namespace is not empty string
+func handleDefaultNamespace(toUpdateBefore map[plugins.ConfigMapID][]ConfigMapUpdate, defaultNamespace string) map[plugins.ConfigMapID][]ConfigMapUpdate {
+	toUpdate := map[plugins.ConfigMapID][]ConfigMapUpdate{}
```
Comment (stevekuznetsov) [inline prow/plugins/updateconfig/updateconfig.go]: Why do we need to duplicate the structure?

Comment (hongkailiu) [reply prow/plugins/updateconfig/updateconfig.go]: Done.

**Comment by hongkailiu** [comment]
/hold cancel

Test added.

File: prow/plugins/updateconfig/updateconfig_test.go [hunk 2]
```diff
@@ -1200,6 +1200,66 @@ func TestUpdateConfig(t *testing.T) {
 	}
 }
 
+func TestFilterChanges(t *testing.T) {
```
Comment (stevekuznetsov) [inline prow/plugins/updateconfig/updateconfig_test.go]: If you want to write `TestFilterChanges` I would expect to see tests for all valid use-cases of `FilterChanges`. I would also like to see the more targeted tests for `handleDefaultNamespace`

Comment (hongkailiu) [reply prow/plugins/updateconfig/updateconfig_test.go]: Done.

**Comment by hongkailiu** [comment]
> I'm fine with this, although we now only test `handleDefaultNamespace` and not `FilterChanges` - would be nice to have tests for that method too, although I understand that covering it in full is quite an additional effort for a bugfix.

Totally true.
Ideally, to mimic your description in the issue, I wanted to get a test for `bootstrapper` where the issue really happened. That would mock the hold process. `FilterChanges` was already 2 steps back from that. And when Steve gave me the `IF` option, here is what we have now. :)

Maybe the point of the current test is:
With `handleDefaultNamespace`, the precondition of the issue does not exist any more. No empty string will be used as `namespace` in the key.

I tried to set it up in `SetDefaults()` function, too. However, the default name comes from `config.yaml` instead of `plugins.yaml` and probably changing the signature of that function is even more involved. So gave this idea up.

**Comment by stevekuznetsov** [comment]
Can you write new test cases for `TestUpdateConfig`?

**Comment by hongkailiu** [comment]
> I don't think this change is complete until a higher-level tests shows me that the config that triggered the bug no longer does.

Added a unit test for `bootstrapper` as discussed.
I had to introduce the function `run` a bit to make logic testable.

File: prow/cmd/config-bootstrapper/main_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+						"bar.yaml": "#bar.yaml",
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		fkc := fake.NewSimpleClientset(tc.existConfigMaps...)
+		tc.configUpdater.SetDefaults()
```
Comment (hongkailiu) [inline prow/cmd/config-bootstrapper/main_test.go]: It has been checked that `SetDefaults()` function has been called (not very explicitly though) before the `run` in the `main` function.

File: prow/cmd/config-bootstrapper/main.go [hunk 4]
```diff
@@ -127,25 +102,52 @@ func main() {
...
+	}
+
+	buildClusterCoreV1Clients, err := o.kubernetes.BuildClusterCoreV1Clients(o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting Kubernetes clients for build cluster.")
+	}
 
-	if errors > 0 {
+	if errors := run(o.sourcePath, configAgent.Config().ProwJobNamespace, pluginAgent.Config().ConfigUpdater, client, buildClusterCoreV1Clients); errors > 0 {
 		logrus.WithField("fail-count", errors).Fatalf("errors occurred during update")
```
Comment (stevekuznetsov) [inline prow/cmd/config-bootstrapper/main.go]: I think you want `len(errors)` for the fail count?

Comment (hongkailiu) [reply prow/cmd/config-bootstrapper/main.go]: `errors` is int, right?
I did not change the type.

**Pull request merged**